lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
-                                             info.depth);
-      builder.create<sv::InterfaceSignalOp>(loc, name, type);
-      break;
+    // Error if a matching parent or companion do not exist.
+    if (parentIDMap.count(bundle.getID()) == 0) {
+      emitCircuitError() << "no parent found with 'id' value '"
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      return false;
     }
-    case ElementInfo::Ground: {
-      auto type = builder.getIntegerType(info.width);
-      builder.create<sv::InterfaceSignalOp>(loc, name, type);
-      break;
+    if (companionIDMap.count(bundle.getID()) == 0) {
+      emitCircuitError() << "no companion found with 'id' value '"
+                         << bundle.getID().getValue().getZExtValue() << "'\n";
+      removalError = true;
+      return false;
     }
-    case ElementInfo::String:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported string type>;");
-      break;
-    case ElementInfo::Boolean:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported boolean type>;");
-      break;
-    case ElementInfo::Integer:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported integer type>;");
-      break;
-    case ElementInfo::Double:
-      builder.create<sv::VerbatimOp>(loc, "// " + name +
-                                              " = <unsupported double type>;");
-      break;
-    case ElementInfo::Error:
-      llvm_unreachable("Shouldn't be here");
-      break;
+
+    // Recursively walk the AugmentedBundleType to generate interfaces and XMRs.
+    // Error out if this returns None.
+    auto iface = traverseBundle(bundle, bundle.getID(),
+                                companionIDMap.lookup(bundle.getID()).name);
+    if (!iface) {
+      removalError = true;
+      return false;
     }
-  }
 
-  interfaces.clear();
-  interfaceMap.clear();
-  interfaceKeys.clear();
+    // Instantiate the interface inside the parent.
+    builder.setInsertionPointToEnd(
+        parentIDMap.lookup(bundle.getID()).second.getBodyBlock());
+    auto symbolName = circuitNamespaceRef->newName(
+        ("__" + companionIDMap.lookup(bundle.getID()).name + "_" +
+         bundle.getDefName().getValue() + "__"));
+    auto instance = builder.create<sv::InterfaceInstanceOp>(
+        getOperation().getLoc(), iface.getValue().getInterfaceType(),
+        companionIDMap.lookup(bundle.getID()).name,
+        builder.getStringAttr(symbolName));
+
+    // If there was no bind file passed in, then we're not supposed to extract
+    // this.  Delete the annotation and continue.  Otherwise, setup the
+    // interface instantiation as a bind.
+    if (!maybeExtractInfo)
+      return true;
```

Comment:
If no bind file is set, is the pass expected to just instantiated the modules inline? Or not do anything at all?

Reply:
That seemed like reasonable behavior to me. However, the SFC will just abort in this situation (assumedly due to the inability of that part of Grand Central to instantiate a SystemVerilog interface in the parent module). I'll leave this in, but make a note that we deviate here and support more behaviors.

I'm fine to change this later, though.

Reply:
I think it's great the way it's implemented here. The comment was more for my understanding than pointing out any flaws :-P